DM-31976: Be more careful with run collection creation and dataset type registration

timj opened the pull request:
By default now configure Butler.transfer_from to not register
dataset types. As part of this reorganization consistency
is now ensured between source butler definitions and target
butler definitions.

## Checklist

- [x] ran Jenkins
- [x] added a release note for user-visible changes to `doc/changes`

timj committed:
Only register dataset types if requested when transferring

By default now configure Butler.transfer_from to not register
dataset types. As part of this reorganization consistency
is now ensured between source butler definitions and target
butler definitions.

timj committed:
Report run collections that are created

Runs are still created inside the transaction but we now
only call registerRun once per run.

timj committed:
Add news fragment

TallJimbo commented on python/lsst/daf/butler/_butler.py:
```diff
@@ -1869,16 +1873,56 @@ def transfer_from(self, source_butler: Butler, source_refs: Iterable[DatasetRef]
+                # search then this process could report failure if not
+                # allowed to register. That is fine. Rerunning will fix
+                # the problem.
+                target_dataset_type = self.registry.getDatasetType(datasetType.name)
```
I think I might prefer a version of this logic that called `registerDatasetType` here when `register_dataset_types=True`, rather than registering only after first identifying it as missing.  Depending on how caching works that could save a database/server query or two (though right now I think our caching is sufficiently aggressive that it doesn't matter in `SqlRegistry`).

Entirely conceivable that that is actually a lot more complex with the way you want to bundle up the error messages, though, and if you've already tried it, ignore me.

timj replied:
Using `registerDatasetType` directly (as was done before) means that we have no way of knowing whether it was newly-created or not. I don't suppose that really matters but it would be nice to let people know. It would help if `registerDatasetType` returned a boolean indicating whether it actually registered anything.

timj replied:
Now that I look at it again, I think that if I just used `registerDatasetType` then that code would only trigger if we are registering dataset types. This would mean that the first we might know about a dataset type not being registered or being inconsistent might be half an hour later when we try to import some data with that dataset type and then rollback the entire transaction. Don't we want to have a hint up front?

TallJimbo replied:
`registerDatasetType` does return a `bool` already.

> This would mean that the first we might know about a dataset type not being registered or being inconsistent might be half an hour later when we try to import some data with that dataset type and then rollback the entire transaction. Don't we want to have a hint up front?

I was thinking that even a half hour earlier is too late, at least for execution butler - we need to get that warning when we make the QG, as per DM-30845.  But yeah, more generally, having an earlier check does seem like a good idea.  How about doing this up front; this still minimizes DB calls and maximizes reuse of checks in daf_butler.
```
if register_dataset_types:
    if butler.registry.registerDatasetType(datasetType):  # let this raise if inconsistent
        log.verbose(...)
else:
    if butler.registry.getDatasetType(name) != datasetType):
        raise ConflictingDefinitionError(...)  # what registerDatasetTypes raises
```
?

Makes me wonder if we should put the else block into an `Registry.expectDatasetType` or an `expect=True` option to `registerDatasetType`, as I bet this isn't the only place that wants this logic.

timj replied:
I rewrote as described.

TallJimbo commented on python/lsst/daf/butler/_butler.py:
```diff
@@ -1924,6 +1927,33 @@ def transfer_from(self, source_butler: Butler, source_refs: Iterable[DatasetRef]
+                missing_run_collections.add(run)
+            elif target_type != CollectionType.RUN:
+                raise ValueError(f"Run collection {run} in source butler is not a RUN in target Butler,"
+                                 f" is {target_type.name}")
```
Moving `DatasetType` registration out of the transaction is _probably_ good even after DM-31956 lands (imminent; just want to write a detailed reply to Andy's comments and then merge the code as-is), because it could involve creating a new table and in some databases (but not SQLite or PostgreSQL!) that's illegal inside a transaction.  But I think moving the collection registration outside the transaction is going to be totally obsoleted by DM-31956:

- There's no scenario where we shouldn't register the collection if it doesn't exist.
- The DM-31956 fix alone is not going to be vulnerable to these kinds of race conditions, and after it merges, the logic here will just open us back up to races.

I think I'd just drop this commit.

timj replied:
What I've ended up doing is retaining the logic here (checking it's not a chain in the target up front seems like a good idea) and just using it to give an informational log message about which runs are likely to be created. This also stops us calling registerRun once per dataset type even if the run hasn't changed.

TallJimbo replied:
Ok, but please pare down the comment, then - this is now just purely informative, as (since DM-31965 has merged) there's nothing to do with race conditions or collection registration not being safe in transactions in play anymore.

timj replied:
I reworked the registerRun stuff to use the new boolean return value to decide whether to report a message.

timj committed:
Fix REPO help string for butler export-calibs

timj committed:
Change dataset type registration logic

Make use of the fact that we know whether the new
dataset type was registered or not.

Also, no longer allow all the dataset types to be
processed before complaining. Raise exception
on first missing dataset type or first inconsistency.

timj committed:
Allow Register.registerCollection to return a boolean

This indicates whether the collection was created or not.

timj committed:
When testing transfer to a CHAIN collection strip other refs

The int->int transfer can randomly end up with clashes if
a second import is attempted, so to remove that randomness
only run the CHAIN test with datasets that we know are in
that collection.  This is needed because we do not
check all the collections up front.
